Source/WebCore/dom/FullscreenManager.cpp

```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
-        m_pendingFullscreenElement = nullptr;
-        return;
+    Vector<RefPtr<Document>> docs { RefPtr { &doc } };
+    while (true) {
+        auto lastDoc = docs.last();
+        if (!lastDoc)
+            continue;
+        ASSERT(lastDoc->fullscreenManager().fullscreenElement());
+        if (!lastDoc->fullscreenManager().isSimpleFullscreenDocument())
+            break;
+        auto parentFrame = lastDoc->frame()->ownerElement();
+        if (!parentFrame || (parentFrame && parentFrame->hasIframeFullscreenFlag()))
+            break;
+        docs.append(&parentFrame->document());
     }
+    return docs;
+}
 
+void FullscreenManager::exitFullscreen()
+{
     INFO_LOG(LOGIDENTIFIER);
 
-    // 3. Let descendants be all the doc's descendant browsing context's documents with a non-empty fullscreen
-    // element stack (if any), ordered so that the child of the doc is last and the document furthest
-    // away from the doc is first.
-    Deque<RefPtr<Document>> descendants;
-    for (AbstractFrame* descendant = frame() ? frame()->tree().traverseNext() : nullptr; descendant; descendant = descendant->tree().traverseNext()) {
-        auto* localFrame = dynamicDowncast<LocalFrame>(descendant);
-        if (!localFrame)
-            continue;
-        if (localFrame->document()->fullscreenManager().fullscreenElement())
-            descendants.prepend(localFrame->document());
-    }
+    auto* currentDoc = &document();
+    auto mode = ExitMode::NoResize;
+    auto exitDocs = documentsToUnfullscreen(*currentDoc);
 
-    // 4. For each descendant in descendants, empty descendant's fullscreen element stack, and queue a
-    // task to fire an event named fullscreenchange with its bubbles attribute set to true on descendant.
-    for (auto& document : descendants) {
-        document->fullscreenManager().clearFullscreenElementStack();
-        addDocumentToFullscreenChangeEventQueue(*document);
+    if (exitDocs.contains(&topDocument()) && topDocument().fullscreenManager().isSimpleFullscreenDocument()) {
+        mode = ExitMode::Resize;
+        currentDoc = &topDocument();
     }
 
-    // 5. While doc is not null, run these substeps:
-    Element* newTop = nullptr;
-    while (currentDoc) {
-        // 1. Pop the top element of doc's fullscreen element stack.
-        currentDoc->fullscreenManager().popFullscreenElementStack();
-
-        //    If doc's fullscreen element stack is non-empty and the element now at the top is either
-        //    not in a document or its node document is not doc, repeat this substep.
-        newTop = currentDoc->fullscreenManager().fullscreenElement();
-        if (newTop && (!newTop->isConnected() || &newTop->document() != currentDoc))
-            continue;
-
-        // 2. Queue a task to fire an event named fullscreenchange with its bubbles attribute set to true
-        // on doc.
+    auto currentDocFullscreenElement = currentDoc->fullscreenManager().fullscreenElement();
```

Comment:
I think we should just call this local `fullscreenElement`. Let’s not use the abbreviation `doc` here.